iree/compiler/Conversion/Common/LinalgBufferizePass.cpp

```diff
@@ -624,19 +945,99 @@ LogicalResult convertInterfaceStoreTensorOp(
   return success();
 }
 
-// Forwards buffer assigned to cast inputs to its outputs.
-LogicalResult convertTensorCastOp(OpBuilder &b,
-                                  WorkgroupMemoryAllocationFn allocationFn,
-                                  tensor::CastOp castOp,
-                                  BlockAndValueMapping &bvm) {
-  Value inputBuffer = bvm.lookup(castOp.source());
-  // Note: tensor.cast isn't suppose to do any data-movements, so we should
-  // never need to allocate and copy data to the result tensor.
-  bvm.map(castOp.dest(), inputBuffer);
+/// Converts a `subtensor_insert` operation to buffers by
+/// - Allocating a buffer for the result (if needed), and copying the
+///   destination value into this buffer.
+/// - Copying the source values into a subview of the result buffer.
+static LogicalResult convertSubTensorInsertOp(OpBuilder &b,
+                                              SubTensorInsertOp op,
+                                              BlockAndValueMapping &bvm,
+                                              BufferizationPlan &plan) {
+  Location loc = op.getLoc();
+  Value result = op.getResult();
+  ShapedType resultType = op.getType();
+  Value resultBuffer = bvm.lookup(result);
+
+  // If `dest` and `result` are not equivalent, need a copy for that.
+  if (!plan.isEquivalent(op.dest(), result)) {
+    Value destBuffer = bvm.lookup(op.dest());
+    b.create<linalg::CopyOp>(loc, destBuffer, resultBuffer);
+  }
+
+  // Copy from the source to the result subview.
+  Value source = op.source();
+  ShapedType sourceType = op.getSourceType();
+  Value sourceBuffer = bvm.lookup(source);
+  SmallVector<OpFoldResult> offsets = op.getMixedOffsets();
+  SmallVector<OpFoldResult> sizes = op.getMixedSizes();
+  SmallVector<OpFoldResult> strides = op.getMixedStrides();
+  MemRefType subViewResultType =
+      (sourceType.getRank() < resultType.getRank()
+           ? memref::SubViewOp::inferRankReducedResultType(
+                 sourceType.getRank(),
+                 resultBuffer.getType().cast<MemRefType>(), offsets, sizes,
+                 strides)
+                 .cast<MemRefType>()
+           : MemRefType());
+  Value subViewOp = createSubviewOp(b, loc, resultBuffer, offsets, sizes,
+                                    strides, subViewResultType);
+  b.create<linalg::CopyOp>(loc, sourceBuffer, subViewOp);
+  return success();
+}
+
+/// Converts a vector.transfer_read op to use memref operands for source.
+static LogicalResult convertVectorTransferReadOp(
+    OpBuilder &b, vector::TransferReadOp transferReadOp,
+    BlockAndValueMapping &bvm) {
+  Value source = transferReadOp.source();
+  if (!source.getType().isa<RankedTensorType>()) return success();
+  Value memref = bvm.lookup(source);
+  transferReadOp.sourceMutable().assign(memref);
```

Comment:
We are creating new ops for others but here we are mutating the original one? Probably better to just create new ones as I think the assumption is old ops aren't touched so we can use information from them anytime.

Reply:
The vector ops we just need to change from using a tensor to using a memref. There is no need to actually create a new op. Actually here new ops are created only when needed (i.e. result changes). For example even `dim` ops above are updated in place. (but shape ops cant be cause the result changes)